widget/scroller.go

```diff
@@ -306,52 +326,59 @@ func (r *scrollContainerRenderer) handleShadowVisibility(offset int, contentSize
 }
 
 func (r *scrollContainerRenderer) updatePosition() {
-	scrollWidth := r.scroll.Size().Width
-	contentWidth := r.scroll.Content.Size().Width
-	scrollHeight := r.scroll.Size().Height
-	contentHeight := r.scroll.Content.Size().Height
-	r.handleAreaVisibility(contentWidth, scrollWidth, r.horizArea)
-	r.handleAreaVisibility(contentHeight, scrollHeight, r.vertArea)
+	scrollSize := r.scroll.Size()
+	contentSize := r.scroll.Content.Size()
 
 	r.scroll.Content.Move(fyne.NewPos(-r.scroll.Offset.X, -r.scroll.Offset.Y))
-	r.handleShadowVisibility(r.scroll.Offset.X, contentWidth, scrollWidth, r.leftShadow, r.rightShadow)
-	r.handleShadowVisibility(r.scroll.Offset.Y, contentHeight, scrollHeight, r.topShadow, r.bottomShadow)
 
-	Renderer(r.vertArea).Layout(r.scroll.size)
-	Renderer(r.horizArea).Layout(r.scroll.size)
+	if r.scroll.Direction != scrollHorizontalOnly {
+		r.handleAreaVisibility(contentSize.Height, scrollSize.Height, r.vertArea)
+		r.handleShadowVisibility(r.scroll.Offset.Y, contentSize.Height, scrollSize.Height, r.topShadow, r.bottomShadow)
+		Renderer(r.vertArea).Layout(r.scroll.size)
+	}
+	if r.scroll.Direction != scrollVerticalOnly {
+		r.handleAreaVisibility(contentSize.Width, scrollSize.Width, r.horizArea)
+		r.handleShadowVisibility(r.scroll.Offset.X, contentSize.Width, scrollSize.Width, r.leftShadow, r.rightShadow)
+		Renderer(r.horizArea).Layout(r.scroll.size)
+	}
 
-	canvas.Refresh(r.vertArea)  // this is required to force the canvas to update, we have no "Redraw()"
-	canvas.Refresh(r.horizArea) // this is required like above but if we are horizontal
+	if r.scroll.Direction != scrollHorizontalOnly {
+		canvas.Refresh(r.vertArea) // this is required to force the canvas to update, we have no "Redraw()"
+	}
+	if r.scroll.Direction != scrollVerticalOnly {
+		canvas.Refresh(r.horizArea) // this is required like above but if we are horizontal
+	}
```

Comment:
Will any test fail if you remove the `if`s here?

Reply:
Added some more tests and now commenting out most 'ifs' and 'switches' will cause test failures, except:
```
if r.scroll.Direction != ScrollHorizontalOnly {
	canvas.Refresh(r.vertArea) // this is required to force the canvas to update, we have no "Redraw()"
}
if r.scroll.Direction != ScrollVerticalOnly {
	canvas.Refresh(r.horizArea) // this is required like above but if we are horizontal
}
```
Since canvas.Refresh is happy to take a nil arg.

Reply:
Okay, `canvas.Refresh(nil)` will probably panic in a real world app with a totally incomprehensible error. So it's better to leave the `if` where they are :)

Reply:
True, it might be worth adding a nil check inside canvas.Refresh